design/one-pager-stacks-security-isolation.md

```diff
@@ -241,13 +277,255 @@ permissionScope: Cluster
+
+We will provide a convenience group - crossplane:master - that binds to crossplane-admin with a
+ ClusterRoleBinding. This mirrors Kubernetes functionality for cluster-admin and system:master. At install time you will
+ then be able to impersonate a group that is bound to crossplane-admin.
+
+Example running a command as crossplane-admin
+```
+kubectl {any crossplane-admin operation} --as-group=crossplane:master
+```
+
+For discoverability of the roles for the environment we will add the following labels:
+ * `rbac.crossplane.io/role-environment: "true"`
+
+Match labels on these roles for auto-aggregation:
+ * `rbac.crossplane.io/aggregate-to-crossplane-admin: "truse"`
+
+#### ClusterRoles for a given app namespace.
+
+For any namespace that is annotated with `crossplane.io/managed-roles: true`, the stack-manager will install and manage
+ the following roles for that namespace that will be properly configured for ClusterRole aggregation as outlined below.
+ These roles should be bound to subjects to grant access to a particular namespace.
+
+ * `crossplane:ns:{ns-name}:admin`
+ * `crossplane:ns:{ns-name}:edit`
+ * `crossplane:ns:{ns-name}:view`
+
+For discoverability of the roles for a given namespace we will add the following labels:
+ * `rbac.crossplane.io/role-namespace: "true"`
+ * `{namespace-name}: "true"`
+
+Match Roles:
+ * `rbac.crossplane.io/aggregate-to-{install-scope}-{role}: "true"`
+ * `{namespace-name}: "true" // to match specific namespace.`
+
+#### ClusterRoles for a given (Cluster)Stack Install
+
+For a stack given stack version the stack-manager will create and or label a unique set of roles in the format
+ stack:{repo-name}:{stack-name}:{version}:{role=system|admin|edit|view}. All of these roles are for internal use only.
+ The system role is bound by the controller for the stack, and will include the set of dependent resources for a given
+ stack. The other (admin,edit,view) will be configured for aggregation to their appropriate namespaced or environment
+ scoped crossplane managed roles.
+
+Consider the following example for a wordpress stack of version 1.1 from wordpressinc, you would have the following set.
+ * stack:wordpressinc:wordpress:1.1:system // Used by wordpress operator - stack resources + dependent resources grants
+ * stack:wordpressinc:wordpress:1.1:admin
+ * stack:wordpressinc:wordpress:1.1:edit
+ * stack:wordpressinc:wordpress:1.1:view
+
+If these were a from a cluster scoped stack, they would be aggregated to the built in environment roles, and if they
+ were from a namespaced stack install they would be aggregated to the corresponding [ClusterRoles for a given app
+ namespace](#clusterroles-for-a-given-app-namespace).
+
+Labels for these roles will be for aggregation purposes:
+ * rbac.crossplane.io/aggregate-to-{install-scope}-{role}: "true"
+ * {namespace-name}: "true" // Only included on namespaced stacks install
```

Comment:
This scheme for the namespace label is a little unintuitive; it would be nice if the reader could tell that the key was a namespace name. It seems we need the namespace in the key (not value) because we want to explicitly whitelist (using labels) the names of multiple namespaces that a `ClusterRole` will be allowed to aggregate up to?

Reply:
Namespaces have a limit of 63 characters, as do labels. I'm not sure about internals of labels, but kubernetes seems to so weird stuff like aggregate-to-view: true, as opposed to aggregate-to: view. I assume maybe the key is indexed and the value is compared, but that's just a guess and what I was following when I did this. `namespace: foo` seems clearer, but not sure if it takes a hit on efficiency, especially considering how many we might have.